Closes #34, #28, #27, #26

sreeharihegden opened the pull request:
# Overview
This pull request makes the following changes:
1. Fix for #34
2. Also fixes #26, #27, and 
3. #28

# Future Directions
N/A

# Notes
N/A

sreeharihegden committed:
Fix for Issue #34: Update gateway() calls to use packaged name.
Also fixes Issues #25, #28.

sreeharihegden committed:
Fix for Issue #34: Update gateway() calls to use packaged name.
Also fixes Issues #25, #28.

sreeharihegden committed:
Move libmexclass install steps to ExternalProject_Add INSTALL_COMMAND.
Also revert previous fix for Issue #25.

sreeharihegden committed:
Move libmexclass install steps to ExternalProject_Add INSTALL_COMMAND.
Also revert previous fix for Issue #25.

sreeharihegden committed:
Update to always use LIBMEXCLASS_DIR as location of LIBMEXCLASS_INSTALL_STEPS_FILE.
Also, updated LIBMEXCLASS_INSTALL_STEPS_FILE to always install to CMAKE_INSTALL_DIR specified from client instead of manipulating CMAKE_INSTALL_PREFIX.

sreeharihegden committed:
Added specific comments to describe the example/CMakeLists.txt CMAKE_INSTALL_DIR value.

sreeharihegden committed:
Corrected libmexclass object file extension during install on Mac to .dylib.
Also, updated example CMakeLists to reflect changes in Kevin's commit: 882e52ee.

sreeharihegden committed:
Corrected libmexclass object file extension during install on Mac to .dylib.
Also, updated example CMakeLists to reflect changes in Kevin's commit: 882e52ee.

kevingurney commented on example/CMakeLists.txt:
```diff
@@ -29,7 +29,13 @@ set(CUSTOM_PROXY_SOURCES
 # Add libmexclass as an external project.
 include(ExternalProject)
-ExternalProject_Add(
+# Clients must use the libmexclass/cpp/libmexclassInstall.cmake to complete the installation steps and package the binaries.
```
Suggested wording: "Clients must use the libmexclass/cpp/libmexclassInstall.cmake..." -> "Clients must use libmexclass/cpp/libmexclassInstall.cmake..."

sreeharihegden replied:
Fixed in commit 309a468.

kevingurney commented on example/CMakeLists.txt:
```diff
@@ -29,7 +29,13 @@ set(CUSTOM_PROXY_SOURCES
 include(ExternalProject)
-ExternalProject_Add(
+# Clients must use the libmexclass/cpp/libmexclassInstall.cmake to complete the installation steps and package the binaries.
+set(LIBMEXCLASS_INSTALL_STEPS_FILE ${LIBMEXCLASS_DIR}/libmexclassInstall.cmake)
```
I prefer the name `install_libmexclass.cmake` to `libmexclassInstall.cmake`.

kevingurney replied:
How will clients access this file if it is coming from `libmexclass`?

sreeharihegden replied:
Install script filename updated in commit 4a2fbf6.

Also, commit 309a468 updated the `ExternalProject_Add` call to get the `libmexclass` source from Git URL. This also resolves Issue #26.

kevingurney commented on example/CMakeLists.txt:
```diff
@@ -47,18 +53,8 @@ ExternalProject_Add(
-# TODO: Only copy libmexclass.so and gateway.mexa64
-install(DIRECTORY ${BINARY_DIR}/ DESTINATION libmexclass)
-install(DIRECTORY ${LIBMEXCLASS_MATLAB_SOURCES} DESTINATION libmexclass)
+install(CODE "MESSAGE(\"Completed install step for ${CLIENT_PROJECT_NAME}.\")")
```
I think we can probably remove this line unless we feel that it isn't clear enough that the external project build has completed successfully.

sreeharihegden replied:
Fixed in commit 309a468 by updating the client install step to take care of copying the packaged libmexclass install to the final destination.

kevingurney commented on libmexclass/cpp/libmexclassInstall.cmake:
```diff
@@ -0,0 +1,31 @@
+if(WIN32)
+	# On Windows, the Visual Studio build adds an additional "Release" folder to place the output binaries.
+	set(ACTUAL_BINARY_DIR "${ACTUAL_BINARY_DIR}/Release/")
+	list(APPEND INSTALL_BINARIES_LIST ${ACTUAL_BINARY_DIR}/gateway.mexw64 ${ACTUAL_BINARY_DIR}/mexclass.dll)	
```
I think we also need to copy over any shared libraries that the client `proxy::Factory` depends on (i.e. `${CUSTOM_PROXY_LINK_LIBRARIES}` to the same directory as the MEX gateway and `libmexclass.dll`

sreeharihegden replied:
Resolved in commit 72a2e7f. See the last `install` steps in the diff for [`libmexclass/cpp/CMakeLists.txt`](https://github.com/mathworks/libmexclass/commit/72a2e7f14cc43fd75dcb298f9841f58238b5db03#diff-7d757216dff7d3606c1796c8225fec917ec2704dd09a8bbc1448f372255e3d90).

sreeharihegden committed:
Updated install script to use MATLAB source from downloaded libmexlass.
Also, removing copying of packaged install to client's target dir.

sreeharihegden committed:
Rename install script libmexclassInstall.cmake to install_libmexclass.cmake.

sreeharihegden committed:
Moved libmexclass install steps to libmexclass/cpp/CMakeLists.txt, and
Updated INSTALL_COMMAND to not use any script.
Also, added install steps in libmexclass to copy CUSTOM_PROXY_LINK_LIBRARIES into the packaged install folder.

sreeharihegden commented:
@kevingurney: Thanks for the earlier review feedback. 
I have resolved all of them. See my comments under each requested change above for the commit details of the fix.

Note that the changes in the PR got simplified as I managed to install `libmexclass` without using any additional script in the `ExternalProject_Add INSTALL_COMMAND`.  See commit 72a2e7f for this change.

- The install steps were moved to [`libmexclass/cpp/CMakeLists.txt`](https://github.com/mathworks/libmexclass/pull/41/files#diff-7d757216dff7d3606c1796c8225fec917ec2704dd09a8bbc1448f372255e3d90).
- `ExternalProject_Add INSTALL_COMMAND` now needs to be simply the CMake command to install.
- The script [`install_libmexclass.cmake`](https://github.com/mathworks/libmexclass/commit/1851089c2f7b827c2f834f1d7e143d6f7c3b51d4) is no longer needed and so removed. 

Changes qualified locally on Windows and Mac.
Also, the latest commit b66f375 passed on Linux CI, see [Actions](https://github.com/mathworks/libmexclass/actions/runs/4350133312).

kevingurney commented:
Thanks for addressing all my feedback @sreeharihegden! This looks great!

sreeharihegden committed:
Updated GIT_TAG to main. Removed extra spaces.

sreeharihegden committed:
Updated example to use GIT_TAG specified with build command.
Also, updated the CI to specify GIT_TAG in the build command.

sreeharihegden committed:
Update CI to run the example on the commit that triggered the worklfow.

kevingurney commented on example/CMakeLists.txt:
```diff
@@ -29,16 +27,24 @@ set(CUSTOM_PROXY_SOURCES
 include(ExternalProject)
+# GIT_TAG for libmexclass source from Git.
+# Note: Clients who don't want to depend on main (which may have breaking changes) can specify a specific commit as argument with cmake command, else source from main branch will be used.
+if(DEFINED GIT_TAG)
```
I would recommend renaming `GIT_TAG` to something like `LIBMEXCLASS_GIT_TAG` since `GIT_TAG` seems a bit generic.

kevingurney replied:
This is a minor suggestion, but I think this could be written a bit more concisely using a CMake cache entry and specifying the default value to `main`.

Reference: https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry

sreeharihegden replied:
Thanks, @kevingurney, for the suggestions. 
Resolved both in the commit 82110aa.

sreeharihegden committed:
Renamed GIT_TAG in CI to LIBMEXCLASS_GIT_TAG.
Also, updated example CMakeLists to use CMake Cache Entry for LIBMEXCLASS_GIT_TAG.
